dev-docs/design/how-lit-html-works.md

```diff
@@ -0,0 +1,426 @@
+```js
+// Pseudo-code of the Prepare phase:
+let preparedTemplate = ChildPart._$getTemplate(counterUi(0));
+
+// Result of the Prepare phase:
+console.log(preparedTemplate.el);
+// <template><span><!--?lit$1234$--></span><button>Increment</button></template>
+
+console.log(preparedTemplate.parts);
+/*
+[
+  { type: 1, index: 0, name: "class" },
+  { type: 2, index: 1 },
+  { type: 1, index: 2, name: "click" }
+]
+*/
+```
+
+The following sections detail what happens in the constructor of the `Template` class
+when preparing a `TemplateResult` for the first time:
+
+#### Join the TemplateResult's template strings array with markers.
+
+We have a list of static HTML strings, but these cannot yet be inserted into a `<template>` element. This pass converts the immutable template strings array into an annotated HTML string where the returned HTML marks the dynamic parts of the HTML with the following annotations:
+
+- Expressions in text position (e.g. `<span>${}</span>`), are marked with a comment node: `<span><!--?lit$random$--></span>`.
+- Expressions in attribute position (e.g. `<p class="${}"></p>`), are marked with a sentinel string: `<p class$lit$="lit$random$"></p>`, and the attribute name has the suffix `$lit$` appended.
+
+The suffix `$lit$` is appended to bound attributes so special-cased attributes (e.g., `style`, `class`, and many SVG attributes) aren't handled by the browser at a point when they contain marker expressions. For instance, IE and Edge parse the `style` attribute value and discard it if it's invalid CSS.
+
+The prepared HTML string for the example's template strings array is:
+
+```js
+getTemplateHtml([
+  '<span class="',
+  '">',
+  '</span><button @click=',
+  '>Increment</button>',
+]);
+
+// Returns the following annotated HTML (spaces added for readability):
+`<span class$lit$="lit$1234$">
+  <?lit$1234$>
+</span>
+<button @click$lit$=lit$1234$>
+  Increment
+</button>`;
+```
+
+Notice that there are three markers that are associated with the example's three dynamic values contained in the `TemplateResult`'s `values` array.
+
+> **Note**
+> In the prepared example HTML string above, the syntax `<?lit$1234$>` represents a [`ProcessingInstruction`](https://developer.mozilla.org/en-US/docs/Web/API/ProcessingInstruction) and is parsed and inserted into the `<template>` element as the comment node: `<!--?lit$1234$-->`. This is an optimization to reduce the number of bytes per marker.
+
+#### Create a `<template>` element
+
+The prepared HTML string is used to set the `innerHTML` of a new `<template>` element. This causes the browser to parse the template HTML. The prepared `<template>` element is assigned to the `Template`'s `el` class field.
+
+This step is safe from XSS vulnerabilities or other malicious input because only the static strings are used and no dynamic values are inserted. Also recall that the `<template>` element only holds HTML, so scripts don't run, styles don't apply, custom elements don't upgrade, etc.
```

Comment:
Dropped!

Reply:
Consider adding:

```markdown
Because the `strings` array consists only of content that was present in source code,
it will not have any dynamic content that could have come from a malicious actor,
unless they already had the ability to run arbitrary code in the page.
```